Doxygen/Docs/man/man3/Cameras.3
.TH "Cameras" 3 "Fri Mar 29 2013" "Version 31337" "HyperGrafx" \" -*- nroff -*-
.ad l
.nh
.SH NAME
Cameras \- 
.PP
The \fBCameras\fP class represents a group of logical cameras for a model view\&. Each camera possesses its own current viewing angle, and an absolute position in space\&.  

.SH SYNOPSIS
.br
.PP
.PP
\fC#include <Cameras\&.hpp>\fP
.PP
Inherits \fBScene\fP\&.
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBCameras\fP (void)"
.br
.RI "\fIDefault constructor\&. \fP"
.ti -1c
.RI "\fB~Cameras\fP (void)"
.br
.RI "\fIDefault destructor\&. \fP"
.ti -1c
.RI "\fBCamera\fP * \fBaddCamera\fP (const std::string &name)"
.br
.RI "\fIaddCamera takes a name for a camera and returns a handle to a newly created camera\&. \fP"
.ti -1c
.RI "void \fBpopCamera\fP (void)"
.br
.RI "\fIpopCamera removes the most recently added \fBCamera\fP from the scene\&. \fP"
.ti -1c
.RI "\fBCamera\fP * \fBnext\fP (void)"
.br
.RI "\fISets the active camera to the next available one in the collection\&. \fP"
.ti -1c
.RI "\fBCamera\fP * \fBprev\fP (void)"
.br
.RI "\fISets the active \fBCamera\fP to the previous available one in the collection\&. \fP"
.ti -1c
.RI "\fBCamera\fP * \fBactive\fP (void) const "
.br
.RI "\fIactive returns the \fBCamera\fP in the collection that is considered 'active'\&. \fP"
.ti -1c
.RI "size_t \fBnumCameras\fP (void) const "
.br
.RI "\fInumCameras fetches the number of \fBCameras\fP in the collection\&. \fP"
.ti -1c
.RI "void \fBidleMotion\fP (void)"
.br
.RI "\fIidleMotion calls the idle method on all child cameras\&. \fP"
.ti -1c
.RI "void \fBresize\fP (int width, int height)"
.br
.RI "\fIresize informs the \fBCameras\fP collection of the new window size\&. \fP"
.ti -1c
.RI "void \fBcalculateViewports\fP (void)"
.br
.RI "\fIFor each \fBCamera\fP in the collection, computes the position and size of that \fBCamera\fP's viewport in a split-screen, single-window configuration\&. \fP"
.ti -1c
.RI "void \fBview\fP (void(*draw_func)(void))"
.br
.RI "\fIview calls the view method on all child cameras, followed by the provided draw function\&. \fP"
.ti -1c
.RI "\fBCamera\fP * \fBobj2Cam\fP (std::list< \fBObject\fP * >::iterator &it)"
.br
.RI "\fIobj2Cam is a gross hack; the function is used as a utility to convert \fBObject\fP pointers to \fBCamera\fP pointers safely\&. \fP"
.ti -1c
.RI "virtual void \fBshader\fP (GLuint gShader)"
.br
.RI "\fISets the Default shader for the scene\&. \fP"
.ti -1c
.RI "GLuint \fBshader\fP (void)"
.br
.RI "\fIRetrieves the handle for the default shader for the scene\&. \fP"
.ti -1c
.RI "\fBObject\fP * \fBaddObject\fP (const std::string &objName, GLuint Object_Shader=0)"
.br
.RI "\fIaddObject creates a new \fBObject\fP with the given name and, optionally, a specified shader and adds it to the \fBScene\fP graph\&. \fP"
.ti -1c
.RI "void \fBdelObject\fP (const std::string &objName)"
.br
.RI "\fIdelObject will remove from the \fBScene\fP graph the object with the given name\&. \fP"
.ti -1c
.RI "void \fBdelObject\fP (void)"
.br
.RI "\fIdelObject with no parameters will delete the first \fBObject\fP in the \fBScene\fP\&. \fP"
.ti -1c
.RI "void \fBpopObject\fP (void)"
.br
.RI "\fIpopObject deletes the last \fBObject\fP in the \fBScene\fP\&. \fP"
.ti -1c
.RI "void \fBdraw\fP (void)"
.br
.ti -1c
.RI "\fBObject\fP * \fBoperator[]\fP (const std::string &objname)"
.br
.ti -1c
.RI "void \fBinsertObject\fP (\fBObject\fP *obj)"
.br
.RI "\fIRegister a created object with the scene graph\&. \fP"
.in -1c
.SS "Protected Member Functions"

.in +1c
.ti -1c
.RI "void \fBdeleteObject\fP (\fBObject\fP *obj)"
.br
.RI "\fIVery seriously delete a child object and free his memory\&. \fP"
.in -1c
.SS "Protected Attributes"

.in +1c
.ti -1c
.RI "std::list< \fBObject\fP * > \fB_list\fP"
.br
.RI "\fIFor the purposes of rapid propagation of scene-graph changes, \fBObject\fP pointers are stored in a regular flat list\&. \fP"
.ti -1c
.RI "std::map< std::string, \fBObject\fP * > \fB_map\fP"
.br
.RI "\fIFor the purposes of accessing named objects quickly, though, objects are also re-stored in an associative map\&. \fP"
.ti -1c
.RI "std::list< \fBObject\fP * >::iterator \fB_currentObj\fP"
.br
.RI "\fIWe keep an iterator on-hand that references what the scene considers to be it's active, current object\&. \fP"
.ti -1c
.RI "GLuint \fB_gShader\fP"
.br
.RI "\fIA handle to a shader program to be used as the default shader for new children objects added to the scene\&. \fP"
.in -1c
.SS "Private Attributes"

.in +1c
.ti -1c
.RI "\fBAngel::vec2\fP \fB_size\fP"
.br
.RI "\fI_size is a simple vec2 (x,y) that contains the size of the screen\&. \fP"
.in -1c
.SH "Detailed Description"
.PP 
The \fBCameras\fP class represents a group of logical cameras for a model view\&. Each camera possesses its own current viewing angle, and an absolute position in space\&. 

\fBAuthor:\fP
.RS 4
John Huston, jhuston@cs.uml.edu 
.RE
.PP
\fBSince:\fP
.RS 4
28 Nov 2012
.RE
.PP
Each \fBCamera\fP possesses its own CTM which can be resent to the GPU at will\&. 
.PP
Definition at line 29 of file Cameras\&.hpp\&.
.SH "Constructor & Destructor Documentation"
.PP 
.SS "Cameras::Cameras (void)"

.PP
Default constructor\&. Nothing special\&. 
.PP
Definition at line 19 of file Cameras\&.cpp\&.
.SS "Cameras::~Cameras (void)"

.PP
Default destructor\&. Nothing special here, either\&. 
.PP
Definition at line 26 of file Cameras\&.cpp\&.
.SH "Member Function Documentation"
.PP 
.SS "\fBCamera\fP * Cameras::active (void) const"

.PP
active returns the \fBCamera\fP in the collection that is considered 'active'\&. \fBReturns:\fP
.RS 4
A pointer to the currently selected, active \fBCamera\fP\&. 
.RE
.PP

.PP
Reimplemented from \fBScene\fP\&.
.PP
Definition at line 88 of file Cameras\&.cpp\&.
.SS "\fBCamera\fP * Cameras::addCamera (const std::string &name)"

.PP
addCamera takes a name for a camera and returns a handle to a newly created camera\&. \fBParameters:\fP
.RS 4
\fIname\fP The name of the new camera to create\&. 
.RE
.PP
\fBReturns:\fP
.RS 4
A Pointer to a newly created \fBCamera\fP object\&. 
.RE
.PP

.PP
Definition at line 35 of file Cameras\&.cpp\&.
.SS "\fBObject\fP * Scene::addObject (const std::string &objName, GLuintshader = \fC0\fP)\fC [inherited]\fP"

.PP
addObject creates a new \fBObject\fP with the given name and, optionally, a specified shader and adds it to the \fBScene\fP graph\&. If no shader is given, a default shader MUST have been specified for the \fBScene\fP prior to the call\&.
.PP
\fBParameters:\fP
.RS 4
\fIobjName\fP The name of the new \fBObject\fP to add\&. 
.br
\fIObject_Shader\fP The shader that should be used to render this object\&. 
.RE
.PP
\fBReturns:\fP
.RS 4
A pointer to the new \fBObject\fP\&. 
.RE
.PP

.PP
Definition at line 77 of file Scene\&.cpp\&.
.SS "void Cameras::calculateViewports (void)"

.PP
For each \fBCamera\fP in the collection, computes the position and size of that \fBCamera\fP's viewport in a split-screen, single-window configuration\&. The \fBCamera\fP object is updated with the new information\&.
.PP
\fBReturns:\fP
.RS 4
void\&.
.PP
void\&. 
.RE
.PP

.PP
Definition at line 143 of file Cameras\&.cpp\&.
.SS "void Scene::deleteObject (\fBObject\fP *obj)\fC [protected]\fP, \fC [inherited]\fP"

.PP
Very seriously delete a child object and free his memory\&. deleteObject is the actual implementation function that will remove an \fBObject\fP from the \fBScene\fP list and \fBScene\fP map, then free the object\&.
.PP
\fBParameters:\fP
.RS 4
\fIobj\fP The object to delete\&.
.br
\fIobj\fP The pointer to the object to free\&. 
.RE
.PP

.PP
Definition at line 129 of file Scene\&.cpp\&.
.SS "void Scene::delObject (const std::string &objName)\fC [inherited]\fP"

.PP
delObject will remove from the \fBScene\fP graph the object with the given name\&. \fBParameters:\fP
.RS 4
\fIobjName\fP Name of the \fBObject\fP to delete\&. 
.RE
.PP

.PP
Definition at line 99 of file Scene\&.cpp\&.
.SS "void Cameras::idleMotion (void)"

.PP
idleMotion calls the idle method on all child cameras\&. Intended to be called during the \fBidle()\fP loop in GLUT\&.
.PP
\fBReturns:\fP
.RS 4
void\&. 
.RE
.PP

.PP
Definition at line 113 of file Cameras\&.cpp\&.
.SS "void Scene::insertObject (\fBObject\fP *obj)\fC [inherited]\fP"

.PP
Register a created object with the scene graph\&. \fBParameters:\fP
.RS 4
\fIname\fP The name of the object (For the associative map), 
.br
\fIobj\fP The \fBObject\fP pointer to add to the scene\&. 
.RE
.PP

.PP
Definition at line 118 of file Scene\&.cpp\&.
.SS "\fBCamera\fP * Cameras::next (void)"

.PP
Sets the active camera to the next available one in the collection\&. Sets the active \fBCamera\fP to the next available one in the collection\&.
.PP
\fBReturns:\fP
.RS 4
A pointer to the newly active \fBCamera\fP\&. 
.RE
.PP

.PP
Reimplemented from \fBScene\fP\&.
.PP
Definition at line 66 of file Cameras\&.cpp\&.
.SS "size_t Cameras::numCameras (void) const"

.PP
numCameras fetches the number of \fBCameras\fP in the collection\&. \fBReturns:\fP
.RS 4
an unsigned integer, the number of \fBCameras\fP in the collection\&. 
.RE
.PP

.PP
Definition at line 103 of file Cameras\&.cpp\&.
.SS "\fBCamera\fP * Cameras::obj2Cam (std::list< \fBObject\fP * >::iterator &it)"

.PP
obj2Cam is a gross hack; the function is used as a utility to convert \fBObject\fP pointers to \fBCamera\fP pointers safely\&. FIXME: Refactor the inheritance here to make this less hacky\&.
.PP
\fBParameters:\fP
.RS 4
\fIit\fP A list<Object*> iterator that points to the \fBObject\fP\&.
.RE
.PP
\fBReturns:\fP
.RS 4
A pointer to a \fBCamera\fP object\&. 
.RE
.PP

.PP
Definition at line 251 of file Cameras\&.cpp\&.
.SS "void Cameras::popCamera (void)"

.PP
popCamera removes the most recently added \fBCamera\fP from the scene\&. \fBReturns:\fP
.RS 4
void\&. 
.RE
.PP

.PP
Definition at line 52 of file Cameras\&.cpp\&.
.SS "\fBCamera\fP * Cameras::prev (void)"

.PP
Sets the active \fBCamera\fP to the previous available one in the collection\&. \fBReturns:\fP
.RS 4
A pointer to the newly active \fBCamera\fP\&. 
.RE
.PP

.PP
Reimplemented from \fBScene\fP\&.
.PP
Definition at line 77 of file Cameras\&.cpp\&.
.SS "void Cameras::resize (intwidth, intheight)"

.PP
resize informs the \fBCameras\fP collection of the new window size\&. Intended to be called from the GLUT main loop\&. This method also invokes \fBCameras::calculateViewports\fP\&.
.PP
\fBParameters:\fP
.RS 4
\fIwidth\fP The new window width\&. 
.br
\fIheight\fP The new window height\&.
.RE
.PP
\fBReturns:\fP
.RS 4
void\&. 
.RE
.PP

.PP
Definition at line 130 of file Cameras\&.cpp\&.
.SS "void Scene::shader (GLuintgShader)\fC [virtual]\fP, \fC [inherited]\fP"

.PP
Sets the Default shader for the scene\&. In the context of inheritance by objects, This sets the shader to use to render the physical object\&.
.PP
\fBParameters:\fP
.RS 4
\fIgShader\fP The GLuint handle to the shader to use\&.
.RE
.PP
\fBReturns:\fP
.RS 4
void\&. 
.RE
.PP

.PP
Reimplemented in \fBObject\fP\&.
.PP
Definition at line 52 of file Scene\&.cpp\&.
.SS "GLuint Scene::shader (void)\fC [inherited]\fP"

.PP
Retrieves the handle for the default shader for the scene\&. In the context of inheritance by objects, This retrieves the shader handle to use to draw the object\&.
.PP
\fBReturns:\fP
.RS 4
A GLuint handle to the shader program\&. 
.RE
.PP

.PP
Reimplemented in \fBObject\fP\&.
.PP
Definition at line 63 of file Scene\&.cpp\&.
.SS "void Cameras::view (void(*)(void)draw_func)"

.PP
view calls the view method on all child cameras, followed by the provided draw function\&. Intended to be called during the \fBdisplay()\fP portion of the GLUT main loop\&.
.PP
\fBview()\fP is intended to 'set up' the object, but not actually draw it\&.
.PP
\fBParameters:\fP
.RS 4
\fIdraw_func\fP A pointer to a function that will actually draw the object\&. 
.RE
.PP

.PP
Definition at line 234 of file Cameras\&.cpp\&.
.SH "Member Data Documentation"
.PP 
.SS "std::map< std::string, \fBObject\fP* > Scene::_map\fC [protected]\fP, \fC [inherited]\fP"

.PP
For the purposes of accessing named objects quickly, though, objects are also re-stored in an associative map\&. It's a little superfluous, but here it is! 
.PP
Definition at line 127 of file Scene\&.hpp\&.

.SH "Author"
.PP 
Generated automatically by Doxygen for HyperGrafx from the source code\&.


src/Classes/Cameras.cpp
/**
 @file Cameras.cpp
 @author John Huston
 @authors John Huston, Nicholas StPierre, Chris Compton
 @date 2012-12-04
 @brief Implementation for the Cameras class, which is a container for Camera objects.
 **/

#include <cmath>
#include <vector>
#include <stdexcept>
#include "Camera.hpp"
#include "Cameras.hpp"
#include "globals.h"

/**
 * Default constructor. Nothing special.
 */
Cameras::Cameras( void ) {
  this->_size = Angel::vec2( 0, 0 );
}

/**
 * Default destructor. Nothing special here, either.
 */
Cameras::~Cameras( void ) {
}

/**
 * addCamera takes a name for a camera and returns a handle to a newly created camera.
 *
 * @param name The name of the new camera to create.
 * @return A Pointer to a newly created Camera object.
 */
Camera *Cameras::addCamera( const std::string &name ) {
  
  Camera *cam = new Camera( name, _gShader );
  
  Scene::insertObject( cam );
  
  calculateViewports();
  return cam;
  //Set active Camera?
  
}

/**
 * popCamera removes the most recently added Camera from the scene.
 *
 * @return void.
 */
void Cameras::popCamera( void ) {
  Scene::popObject();
  calculateViewports();
}

//DelCamera( const std::string &name );
//DelCamera( void );
//DestroyCamera( void );

/**
 * Sets the active Camera to the next available one in the collection.
 *
 * @return A pointer to the newly active Camera.
 */
Camera *Cameras::next( void ) {
  Scene::next();
  active()->send( Camera::CTM );
  return active();
}

/**
 * Sets the active Camera to the previous available one in the collection.
 *
 * @return A pointer to the newly active Camera.
 */
Camera *Cameras::prev( void ) {
  Scene::prev();
  active()->send( Camera::CTM );
  return active();
}

/**
 * active returns the Camera in the collection that is considered 'active'.
 *
 * @return A pointer to the currently selected, active Camera.
 */
Camera *Cameras::active( void ) const {
  
  // Use the base method to retrieve the active object.
  Object *camptr = Scene::active();
  
  // Convert it to a Camera pointer.
  return dynamic_cast< Camera* >( camptr );
  
}

/**
 * numCameras fetches the number of Cameras in the collection.
 *
 * @return an unsigned integer, the number of Cameras in the collection.
 */
size_t Cameras::numCameras( void ) const {
  return _list.size();
}

/**
 * idleMotion calls the idle method on all child cameras.
 * Intended to be called during the idle() loop in GLUT.
 *
 * @return void.
 */
void Cameras::idleMotion( void ) {
  std::list< Object* >::iterator it;
  for ( it = _list.begin(); it != _list.end(); ++it ) {
    obj2Cam( it )->idle();
  }
}

/**
 * resize informs the Cameras collection of the new window size.
 * Intended to be called from the GLUT main loop.
 * This method also invokes Cameras::calculateViewports.
 *
 * @param width The new window width.
 * @param height The new window height.
 *
 * @return void.
 */
void Cameras::resize( int width, int height ) {
  this->_size = Angel::vec2( width, height );
  if ( DEBUG )
    std::cerr << "Setting Cameras WxH: " << width << "x" << height << "\n";
  calculateViewports();
}

/**
 * For each Camera in the collection, computes the position and size of that
 * Camera's viewport in a split-screen, single-window configuration.
 *
 * @return void.
 */
void Cameras::calculateViewports( void ) {
  
  unsigned Width = _size.x;
  unsigned Height = _size.y;
  
  /*
   * This function is a little confusing.
   * Ultimately, we see how many virtual cameras we have,
   * then we compute how many "rows" and "columns" of virtual viewports
   * we will need to accommodate that many.
   *
   * After those calculations, we run through a loop that goes
   * row-by-row and then column-by-column computing the actual
   * x,y offsets, height and widths of each viewport.
   */

  /* Let's not try to distribute Viewports if the window system
   hasn't been initialized yet, i.e, if resize() has not yet
   been called. */
  if ( !_size.x || !_size.y ) return;
  
  // How many cameras do we have?
  size_t numCameras = Cameras::numCameras();
  
  // Let's not try to resize zero cameras.
  if ( numCameras == 0 ) return;
  
  // How many virtual rows will we need to display this many?
  size_t numRows = ceil( sqrt( numCameras ) );
  
  // How many columns will we need (average) in general to display this many?
  double numCols = (double) numCameras / (double) numRows;
  size_t numMaxCols = ceil( numCols );
  size_t numMinCols = floor( numCols );
  
  // How many rows do we need to draw with MinCols?
  // (By extension: drawMaxRows = (numRows-drawMinRows))
  size_t drawMinRows =
      (numCameras % numRows) ?
          ((int) (numCameras - (numRows * numMaxCols))
           / (int) (numMinCols - numMaxCols)) :
          0;
  
  std::list< Object * >::iterator it = _list.begin();
  
  /* This is confusing as hell. Good luck! */
  for ( size_t allocHeight = 0, row = 0; (row < numRows) || (it != _list.end());
      ++row ) {
    size_t myWidth;
    size_t myHeight;
    size_t allocWidth = 0;
    size_t colsThisRow = (row < drawMinRows) ? numMinCols : numMaxCols;
    if ( 0 )
      fprintf( stderr, "Row: %lu; Columns this row: %lu\n", row, colsThisRow );
    
    for ( size_t col = 0; col < colsThisRow; ++col, ++it ) {
      // Is this the last column? Use the remaining width.
      if ( col + 1 == colsThisRow ) myWidth = (Width) - allocWidth;
      else myWidth = (Width) / colsThisRow;
      // Is this the last row? Use the remaining height.
      if ( row + 1 == numRows ) myHeight = (Height) - allocHeight;
      else myHeight = (Height) / numRows;
      
      // Tell this camera his new viewport.
      // height looks a little goofy because we are allocating height
      // from the top of the coordinate system and working down,
      // so we have to take the complement.
      obj2Cam( it )->viewport( allocWidth,
                               ((Height) - (allocHeight + myHeight)), myWidth,
                               myHeight );
      if ( 0 )
        fprintf( stderr, "Camera: (%lu x %lu) @ (%lu,%lu)\n", myWidth, myHeight,
                 allocWidth, ((Height) - (allocHeight + myHeight)) );
      
      // Increment our allocated width counter.
      allocWidth += myWidth;
    }
    // Increment our allocated height counter.
    allocHeight += ((Height) / numRows);
  }
}

/**
 * view calls the view method on all child cameras, followed by
 * the provided draw function. Intended to be called during the
 * display() portion of the GLUT main loop.
 *
 * view() is intended to "set up" the object, but not actually draw it.
 *
 * @param draw_func A pointer to a function that will actually draw the object.
 */
void Cameras::view( void (*draw_func)( void ) ) {
  std::list< Object* >::iterator it;
  for ( it = _list.begin(); it != _list.end(); ++it ) {
    obj2Cam( it )->view();
    (*draw_func)();
  }
}

/**
 * obj2Cam is a gross hack; the function is used as a utility to
 * convert Object pointers to Camera pointers safely.
 * FIXME: Refactor the inheritance here to make this less hacky.
 *
 * @param it A list<Object*> iterator that points to the Object.
 *
 * @return A pointer to a Camera object.
 */
Camera *Cameras::obj2Cam( std::list< Object* >::iterator &it ) {
  return dynamic_cast< Camera* >( *it );
}
